imagecraft/pack/gptutil.py

```diff
@@ -0,0 +1,256 @@
+    run(
+        "sfdisk",
+        imagepath,
+        input=stdin_lines,
+    )
+
+
+def create_empty_gpt_image(
+    imagepath: Path,
+    sector_size: int,
+    layout: Volume,
+) -> None:
+    """Create a zeroed image file with a GPT partition table, but no filesystems or data.
+
+    :param imagepath: Path to image file.
+    :param sector_size: Sector size in bytes.
+    :param layout: Disk layout to create.
+    """
+    # Determine necessary image size, and reserve space.
+    # "The default start offset for the first partition is 1 MiB", per `man sfdisk`,
+    # plus one more MiB so we don't overwrite the backup GPT header at the end of the
+    # disk.
+    image_bytes = diskutil.MIB * 2
+    for structure_item in layout.structure:
+        image_bytes += structure_item.size
+    disk_size = diskutil.DiskSize(bytesize=image_bytes, sector_size=sector_size)
+    emit.trace(f"Creating {image_bytes} byte empty image")
+
+    diskutil.create_zero_image(imagepath=imagepath, disk_size=disk_size)
+
+    # Create partition table
+    _create_gpt_layout(
+        imagepath=imagepath,
+        sector_size=sector_size,
+        layout=layout,
+    )
+
+
+def _get_partition_table(imagepath: Path) -> dict[str, Any]:
+    """Return a dict representing the complete partition table.
+
+    :raises CalledProcessError: If sfdisk fails.
+    """
+    return json.loads(run("sfdisk", "--json", imagepath).stdout)["partitiontable"]  # type: ignore[no-any-return]
+
+
+def primary_gpt_sector_count(*, imagepath: Path) -> int:
+    """Extract which sector the data starts.
+
+    :param imagepath: Path to image file.
+    :raises CalledProcessError: If sfdisk fails.
+    """
+    # This is the start of the usable GPT sectors after the header
+    # so we know every sector before this is part of the header.
+    return int(_get_partition_table(imagepath)["firstlba"])
+
+
+def extract_primary_gpt(*, imagepath: Path, sector_size: int, headerpath: Path) -> None:
+    """Extract the header sectors.
+
+    :param imagepath: Path to image file.
+    :param sector_size: Sector size in bytes.
+    :param headerpath: Path to header file for writing.
+    :raises CalledProcessError: If sfdisk or dd fails.
+    """
+    count = primary_gpt_sector_count(imagepath=imagepath)
+
+    # Sectors start at zero, so the total is the count to copy
+    run(
+        "dd",
+        f"if={str(imagepath)}",
+        f"of={str(headerpath)}",
+        f"bs={sector_size}",
+        f"count={count}",
+    )
+
+
+def backup_gpt_sector_start(*, imagepath: Path) -> int:
+    """Extract which sector the data starts.
+
+    :param imagepath: Path to image file.
+    :raises CalledProcessError: If sfdisk fails.
+    """
+    # This is the end of the usable GPT sectors before the footer
+    # so we know every sector after this is the footer.
+    return int(_get_partition_table(imagepath)["lastlba"]) + 1
+
+
+def extract_backup_gpt(*, imagepath: Path, sector_size: int, footerpath: Path) -> None:
```

Comment:
It looks like this one and  `extract_primary_gpt` (and so  `backup_gpt_sector_start` and `primary_gpt_sector_count`) are not used. I think we can remove them.